### PR TITLE
UCP: error handling implementation

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
see #755

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1511/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3432/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1557/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3475/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1563/ for details.

File: src/ucp/core/ucp_ep.c [hunk 1]
```diff
@@ -212,12 +219,23 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker,
         goto err_destroy_ep;
     }
 
+    /* Setup error handler */
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: missed it in previous PR, but there is a problem with ucp_ep_config(ep)->err_mode. Need to configure the err mode inside the "key" field of ep_config **before** filling the rest of the configuration. Consider 2 endpoints which have same configuration, except err handling. they will end up using same config struct and override each other settings.

File: src/ucp/core/ucp_ep.h [hunk 2]
```diff
@@ -135,6 +135,8 @@ typedef struct ucp_ep_config {
 
     /* Error handling mode */
     ucp_err_handling_mode_t    err_mode;
+    /* Status of the endpoint */
+    ucs_status_t               status;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.h]: this should not be here - see comment about err_mode. what if u have 2 endpoints with same config, on of them fails and other one is still OK?

File: src/ucp/core/ucp_ep.c [hunk 3]
```diff
@@ -447,6 +468,13 @@ static void ucp_ep_disconnected(ucp_request_t *req)
 {
     ucp_ep_h ep = req->send.ep;
 
+    /* Delay cleanup of failed EPs until worker destroy */
+    if (ucp_ep_config(ep)->status != UCS_OK) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: would be better to check ep->flags here, if needed define more flags for err handling

File: src/ucp/core/ucp_mm.h [hunk 4]
```diff
@@ -59,8 +59,9 @@ void ucp_rkey_resolve_inner(ucp_rkey_h rkey, ucp_ep_h ep);
 
 #define UCP_RKEY_RESOLVE(_rkey, _ep, _op_type) \
     ({ \
-        ucs_status_t status = UCS_OK; \
-        if (ucs_unlikely((_ep)->cfg_index != (_rkey)->cache.ep_cfg_index)) { \
+        ucs_status_t status = ucp_ep_config(_ep)->status; \
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.h]: why do we need to check err status here? isn't it enough to put dummy func pointers for ucp put/get/atomic?

File: src/ucp/core/ucp_worker.c [hunk 5]
```diff
@@ -218,10 +218,73 @@ ucp_to_uct_wakeup_events(ucp_wakeup_event_t ucp_events)
...
+    ucp_ep_h           ucp_ep           = NULL;
+    uint64_t           dest_uuid        = 0ul;
+    ucp_ep_h           ucp_ep_iter;
+    khiter_t           ucp_ep_errh_iter;
+    ucp_err_handler_t  err_handler;
+    ucp_rsc_index_t    lane, n_lanes, failed_lane;
+
+    /* TODO: need to optimize uct_ep -> ucp_ep lookup */
+    kh_foreach(&worker->ep_hash, dest_uuid, ucp_ep_iter, {
+        for (lane = 0; (lane < ucp_ep_num_lanes(ucp_ep_iter)); ++lane) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: no need for extra ( )

File: src/ucp/core/ucp_worker.c [hunk 6]
```diff
@@ -218,10 +218,73 @@ ucp_to_uct_wakeup_events(ucp_wakeup_event_t ucp_events)
...
+            if ((uct_ep == ucp_ep_iter->uct_eps[lane]) && dest_uuid) {
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: why need to check that dest_uuid!=0?

File: src/ucp/core/ucp_worker.c [hunk 7]
```diff
@@ -218,10 +218,73 @@ ucp_to_uct_wakeup_events(ucp_wakeup_event_t ucp_events)
...
+    });
+
+    ucs_fatal("No uct_ep_h %p associated with ucp_ep_h on ucp_worker_h %p",
+              uct_ep, worker);
+    return;
+
+found_ucp_ep:
+    ucp_ep_config(ucp_ep)->status = status;
+
+/* Destroy all lanes excepting failed one since ucp_ep becomes unusable as well */
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: indentation

File: src/ucp/core/ucp_worker.c [hunk 8]
```diff
@@ -218,10 +218,73 @@ ucp_to_uct_wakeup_events(ucp_wakeup_event_t ucp_events)
...
+
+    /* Move failed lane at 0 index */
+    if (failed_lane != 0) {
+        ucp_ep->uct_eps[0] = ucp_ep->uct_eps[failed_lane];
+        ucp_ep->uct_eps[failed_lane] = NULL;
+    }
+
+    /* Redirect all lanes to failed one */
+    ucp_ep->am_lane                        =
+    ucp_ep_config(ucp_ep)->key.am_lane     =
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: this cannot be done - breaks all other endpoints using same config.
need to create a new configuration key and allocate index using ucp_worker_get_ep_config()

File: test/gtest/ucp/test_ucp_error_handling.cc [hunk 9]
```diff
@@ -9,14 +9,112 @@
 
 class test_ucp_error_handling : public test_ucp_tag {
 public:
+    test_ucp_error_handling() : err_cntr(0ul), err_status(UCS_OK) {}
```
Comment (yosefe) [inline test/gtest/ucp/test_ucp_error_handling.cc]: need to make a test with 2 pairs, check that after 1 pair fails the other one still works (to expose the config bug above)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3481/ for details (Mellanox internal link).

File: src/ucp/core/ucp_mm.h (hunk 4 above)
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_mm.h]: We do not have func pointers in ucp for API (like in UCT). this is needed to invalidate cached lane

File: src/ucp/core/ucp_worker.c (hunk 6 above)
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_worker.c]: To fix compiler's warning/error "not used variable"

File: src/ucp/core/ucp_mm.h (hunk 4 above)
Comment (yosefe) [inline src/ucp/core/ucp_mm.h]: wouldn't it be enough to update ep->cfg_index to invalidate the cache?

File: src/ucp/core/ucp_worker.c (hunk 6 above)
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: pls use UCS_V_UNUSED or kh_foreach_value

File: src/ucp/core/ucp_mm.h (hunk 4 above)
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_mm.h]: then it would break whole configuration of the endpoint, it's hard for me to predict the impact, event for failed EP :)

Comment (yosefe) [inline src/ucp/core/ucp_mm.h]: why? the new cfg_index should point to a configuration with the new set of lanes (where the failed lane at index 0 and rest are removed), so rkey lookup should work. then it would call the put/get/amo operations of failed lane, and return the status..

Comment (evgeny-leksikov) [inline src/ucp/core/ucp_mm.h]: it should work, thanks!

File: src/ucp/core/ucp_worker.c (hunk 8 above)
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_worker.c]: @yosefe I removed status from the key and use flags instead of status on disconnection as you recommended. But keeping status in the key would fix this issue, all failed EPs would have the own config with first failed uct lane.
May be to return that?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1568/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3486/ for details (Mellanox internal link).

File: src/ucp/core/ucp_ep.c [hunk 10]
```diff
@@ -219,26 +228,27 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker,
         goto err_destroy_ep;
     }
 
+    if (err_mode == UCP_ERR_HANDLING_MODE_PEER) {
+        /* Disable RNDV */
+        ucp_ep_config(ep)->rndv.am_thresh = SIZE_MAX;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: this should be done inside ucp_ep_config_init() - outside this function, the config should not be modified.

File: src/ucp/core/ucp_ep.c [hunk 11]
```diff
@@ -219,26 +228,27 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker,
...
+    }
+
     /* Setup error handler */
-    if (params->field_mask & UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE) {
-        ucp_ep_config(ep)->err_mode = params->err_mode;
-        if (params->err_mode == UCP_ERR_HANDLING_MODE_PEER) {
-            /* Disable RNDV */
-            ucp_ep_config(ep)->rndv.am_thresh = SIZE_MAX;
-        }
+    hash_it = kh_put(ucp_ep_errh_hash, &worker->ep_errh_hash, (uintptr_t)ep,
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: need to add to hash only if err handler is passed

File: src/ucp/core/ucp_ep.c [hunk 12]
```diff
@@ -39,13 +39,14 @@ void ucp_ep_config_key_reset(ucp_ep_config_key_t *key)
     key->rndv_lane        = UCP_NULL_LANE;
     key->wireup_lane      = UCP_NULL_LANE;
     key->reachable_md_map = 0;
+    key->err_mode         = UCP_ERR_HANDLING_MODE_NONE;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: need to update ucp_ep_config_is_equal()

File: src/ucp/core/ucp_worker.c [hunk 13]
```diff
@@ -246,17 +246,17 @@ uct_to_ucp_error_proxy(void *arg, uct_ep_h uct_ep, ucs_status_t status)
...
         if ((lane != failed_lane) && ucp_ep->uct_eps[lane]) {
             uct_ep_destroy(ucp_ep->uct_eps[lane]);
             ucp_ep->uct_eps[lane] = NULL;
-            ucp_ep_config(ucp_ep)->key.num_lanes--;
         }
     }
-    ucs_assert(ucp_ep_config(ucp_ep)->key.num_lanes == 1);
+
+    ucp_ep_config_key_t key = ucp_ep_config(ucp_ep)->key;
+    ucp_ep->cfg_index = ucp_worker_get_ep_config(worker, &key);
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: this is not right - need to fill the key **before** calling this function, otherwise it's pointless

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1573/ for details.

File: src/ucp/core/ucp_mm.h [hunk 14]
```diff
@@ -59,8 +59,9 @@ void ucp_rkey_resolve_inner(ucp_rkey_h rkey, ucp_ep_h ep);
 
 #define UCP_RKEY_RESOLVE(_rkey, _ep, _op_type) \
     ({ \
-        ucs_status_t status = UCS_OK; \
-        if (ucs_unlikely((_ep)->cfg_index != (_rkey)->cache.ep_cfg_index)) { \
+        ucs_status_t status = ucp_ep_config(_ep)->key.status; \
```
Comment (yosefe) [inline src/ucp/core/ucp_mm.h]: didn't we discuss relying on put/get ops to return the status in case of failure?

**Comment by yosefe** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3504/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1579/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1580/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3505/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1585/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3510/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1586/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3511/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
@evgeny-leksikov i've pushed my work to this PR,
TODO:
 - fix and test dcx and rcx transports error handling (there is a problem with counting completions) - need to be in separate PR (incl. uct multipair test)
 - remove debug prints
 - add multi-pair support to ucp_test
 - test different operations on UCP
 - test when some operations are successful and others have failed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1588/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3513/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1655/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3573/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1664/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3581/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1697/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3629/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3633/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1755/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1756/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3693/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3694/ for details (Mellanox internal link).

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3697/ for details (Mellanox internal link).

File: src/uct/ib/dc/base/dc_iface.c [hunk 15]
```diff
@@ -490,6 +490,7 @@ void uct_dc_handle_failure(uct_ib_iface_t *ib_iface, uint32_t qp_num)
         uct_dc_iface_dci_put(iface, dci);
         ucs_assert_always(ep->dci == UCT_DC_EP_NO_DCI);
     }
+    iface->super.tx.cq_available++;
```
Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: this is correct fix, pls add documentation - that it is because poll_cqe for mlx5 returns NULL in case of failure and the cq_avaialble is not updated for the **error cqe**

Comment (yosefe) [inline src/uct/ib/dc/base/dc_iface.c]: BTW this can go to separate PR + test on UCT level

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1769/ for details.

**Comment by evgeny-leksikov** [comment]
@yosefe I reworked a bit cq_avalable counting for DC/verbs because of common handle failure code.
As we agreed this PR should be merged without splitting.
Should I rebase it on master if everything is OK?(there is a merge commit with conflicts resolving)

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3707/ for details (Mellanox internal link).

File: src/ucp/core/ucp_ep.c [hunk 16]
```diff
@@ -209,19 +218,32 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker,
...
 
-    if (params->field_mask & UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE) {
+    /* Setup error handler */
+    if (params->field_mask & UCP_EP_PARAM_FIELD_ERR_HANDLER) {
+        hash_it = kh_put(ucp_ep_errh_hash, &worker->ep_errh_hash, (uintptr_t)ep,
+                         &hash_extra_status);
+        if (ucs_unlikely(hash_it == kh_end(&worker->ep_errh_hash))) {
+            ucs_error("Hash failed on setup error handler of endpoint %p with status %d ",
+                      ep, hash_extra_status);
+            status = UCS_ERR_NO_RESOURCE;
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: UCS_ERR_NO_MEMORY

File: src/ucp/core/ucp_ep.c [hunk 17]
```diff
@@ -209,19 +218,32 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker,
...
+            goto err_destroy_ep;
+        }
+        kh_value(&worker->ep_errh_hash, hash_it) = params->err_handler;
+
         ucp_ep_config(ep)->err_mode = params->err_mode;
+    }
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: the line above should be removed, since err_mode is part of the key

File: src/ucp/core/ucp_ep.c [hunk 18]
```diff
@@ -209,19 +218,32 @@ ucs_status_t ucp_ep_create(ucp_worker_h worker,
...
+
+    if (params->field_mask & UCP_EP_PARAM_FIELD_ERR_HANDLING_MODE) {
```
Comment (yosefe) [inline src/ucp/core/ucp_ep.c]: need to move this code to ucp_ep_config_init() to set rndv_thresh based on key->err_mode

File: src/ucp/core/ucp_worker.c [hunk 19]
```diff
@@ -226,10 +226,87 @@ ucp_to_uct_wakeup_events(ucp_wakeup_event_t ucp_events)
...
+    }
+
+    /* Redirect all lanes to failed one */
+    ucp_ep_config_key_t key = ucp_ep_config(ucp_ep)->key;
+    key.am_lane            = 0;
+    key.wireup_lane        = 0;
+    key.rndv_lane          = 0;
+    key.amo_lanes[0]       = 0;
+    key.rma_lanes[0]       = 0;
+    key.lanes[0].rsc_index = UCP_NULL_RESOURCE;
```
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: need to check if operations which do registration (e.g zcopy) fail correctly on a failed ep, and there is no problem with memory registration or cleanup of memory registration

File: src/ucp/wireup/wireup.c [hunk 20]
```diff
@@ -190,8 +191,15 @@ static void ucp_wireup_process_request(ucp_worker_h worker, const ucp_wireup_msg
         }
     }
 
+    /* Use default parameters to initialize local lanes.
+     * This might not always work - need to pass some parameters in the wireup
```
Comment (yosefe) [inline src/ucp/wireup/wireup.c]: pls add to TODO / or to this PR:
pass error handling mode in wireup message, in case ep is created internally, as a result of peer request, before the user creates it himself. in this case ep will have err mode "NONE" even though user could have requested err mode "PEER"

File: test/gtest/ucp/ucp_test.cc [hunk 21]
```diff
@@ -114,6 +114,9 @@ void ucp_test::disconnect(const entity& entity) {
         entity.flush_worker(i);
         void *dreq = entity.disconnect_nb(i);
         if (!UCS_PTR_IS_PTR(dreq)) {
+            if (UCS_ERR_ENDPOINT_TIMEOUT == UCS_PTR_STATUS(dreq)) {
```
Comment (yosefe) [inline test/gtest/ucp/ucp_test.cc]: why is this needed? does it happen only for err handling tests?
need to make sure we don't ignore errors on all other tests

File: src/ucp/core/ucp_worker.c (hunk 19 above)
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_worker.c]: we have test with RMA ops after EP failure, failed RMA operation is a local error, it should be handled respectively.
I'll add a test with failing of 'zcopy' message.

File: src/ucp/wireup/wireup.c (hunk 20 above)
Comment (evgeny-leksikov) [inline src/ucp/wireup/wireup.c]: I'm not sure that increase of error handling mode greater than requested is correct. I think we need to reconfigure existing EP if it was created with default config/key on wire up but requested something else.

Comment (evgeny-leksikov) [inline src/ucp/wireup/wireup.c]: Added in TODO here: https://github.com/openucx/ucx/pull/1461/files#diff-8fd5f37669b4533d1e9389ff6839dcccR210

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/1789/ for details.

File: src/ucp/core/ucp_worker.c (hunk 19 above)
Comment (yosefe) [inline src/ucp/core/ucp_worker.c]: @evgeny-leksikov do you plan to add it to another PR?
can you please update https://github.com/openucx/ucx/issues/755 with a TODO ?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3726/ for details (Mellanox internal link).

File: src/ucp/core/ucp_worker.c (hunk 19 above)
Comment (evgeny-leksikov) [inline src/ucp/core/ucp_worker.c]: @yosefe let's do it separately. #755 is updated

**Comment by evgeny-leksikov** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3731/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3734/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@yosefe - this pr requires two reviews before merge
